_posts/2022-01-31-how-we-ship-ios-libraries.md

```diff
@@ -0,0 +1,333 @@
+We will manually modify that entry next.
+Do the same for the `Release` configuration.
+
+Once that's done, save your project and go back to your project directory.
+We will modify the project configuration to have Xcode look for the library based on the target it is building for[^2].
+Open up `ShippingRust.xcodeproj/project.pbxproj` in a text editor,
+then search for the first line with `"LIBRARY_SEARCH_PATHS[sdk=driverkit*]"`.
+It should be in a section saying `/* Debug */`.
+Remove that line and add 3 new ones:
+
+```
+"LIBRARY_SEARCH_PATHS[sdk=iphoneos*][arch=arm64]" = "$(PROJECT_DIR)/target/aarch64-apple-ios/debug";
+"LIBRARY_SEARCH_PATHS[sdk=iphonesimulator*][arch=arm64]" = "$(PROJECT_DIR)/target/aarch64-apple-ios-sim/debug";
+"LIBRARY_SEARCH_PATHS[sdk=iphonesimulator*][arch=x86_64]" = "$(PROJECT_DIR)/target/x86_64-apple-ios/debug";
+```
+
+Look for the next line with `"LIBRARY_SEARCH_PATHS[sdk=driverkit*]"`, now in a `/* Release */` section and replace it with:
+
+```
+"LIBRARY_SEARCH_PATHS[sdk=iphoneos*][arch=arm64]" = "$(PROJECT_DIR)/target/aarch64-apple-ios/release";
+"LIBRARY_SEARCH_PATHS[sdk=iphonesimulator*][arch=arm64]" = "$(PROJECT_DIR)/target/aarch64-apple-ios-sim/release";
+"LIBRARY_SEARCH_PATHS[sdk=iphonesimulator*][arch=x86_64]" = "$(PROJECT_DIR)/target/x86_64-apple-ios/release";
+```
+
+Save the file and return back to Xcode.
+If you didn't make any typos Xcode should still be able to open your project.
+In the settings you will find the library search paths as we've just defined them.
+
+Next we need to teach Xcode how to compile Rust code.
+Once again go to your target settings, selecting the "Build Phases" tab again.
+
+There add a new "Run Script" phase, give it a name (double-click the "Run Script" section header),
+and set the command to:
+
+```
+bash ${PROJECT_DIR}/bin/compile-library.sh shipping-rust-ffi $buildvariant
+```
+
+The `compile-library.sh` script is going to do the heavy lifting.
+The first argument is the crate name we want to compile, the second is the build variant to select.
+This is not yet defined, so let's do it first.
+
+Go to the "Build Settings" tab and click the `+` button to add a new "User-Defined Setting".
+Give it the name `buildvariant` and choose a value based on the build variant: `debug` for `Debug` and `release` for `Release`.
+
+Now we're ready to write the `compile-library.sh` script.
+Create a new directory `bin` in your project directory.
+In there create the file with the following content:
+
+```bash
+#!/usr/bin/env bash
+
+# This should be invoked from inside xcode, not manually
+if [ "$#" -ne 2 ]
+then
+    echo "Usage (note: only call inside xcode!):"
+    echo "Args: $*"
+    echo "path/to/build-scripts/xc-universal-binary.sh <FFI_TARGET> <buildvariant>"
+    exit 1
+fi
+
+# what to pass to cargo build -p, e.g. your_lib_ffi
+FFI_TARGET=$1
+# buildvariant from our xcconfigs
+BUILDVARIANT=$2
+
+RELFLAG=
+if [[ "$BUILDVARIANT" != "debug" ]]; then
+    RELFLAG=--release
```

Comment:
This indents 4 spaces, but the below code indents 2. I mention it because maybe it indicates a missing line or copy/paste error.